### DISCUSSION TITLE
About React ReArch

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#144] [General]
The three main lifecycle events of a React Component are `componentDidMount`, `componentDidUpdate` and `componentWillUnmount`.

I suppose that I can  discard `componentDidMount`. But what about `componentDidUpdate`? Should I advice ReArch about this event. React state gets substituted by ReArch state, but what about React properties?

https://legacy.reactjs.org/docs/react-component.html#componentdidupdate

**Comment by busslina** [comment]
Okay, I think it can be ignored too because it has attached the `render` method which is executed just before `componentDidUpdate`. So the question would be then: Is it safe to use `properties` from the render/build method?

**Reply by GregoryConrad** [reply]
I’m not too familiar with React so you’ll have to play around with that a bit haha

**Comment by busslina** [comment]
In the ReArch context what does mean that a Capsule is idempotent?

I'm just trying to replicate a similar mechanism to "Flutter Widget dirty".

**Reply by GregoryConrad** [reply]
A capsule is idempotent if it _does not_ register any side effects. This makes a capsule's state entirely reproducible given the state of the rest of the container, and can enable it to be garbage collected (removed from the container) when it has no demand (no direct or indirect dependents that register a side effect, which also includes listeners).

And the markNeedsBuild should be replaceable simply with a call to setState over in React; just call it via onNextUpdate.

**Reply by busslina** [reply]
I think your suggestion is mostly what I do now (via `rebuild()`) (`forceUpdate` acts similar to `setState`).

```dart
@override
  T call<T>(Capsule<T> capsule) {
    final dispose = container.onNextUpdate(capsule, api.rebuild);
    api.component.dependencyDisposers.add(dispose);
    return container.read(capsule);
  }
```
```dart
@override
  void rebuild([
    void Function(void Function() cancelRebuild)? sideEffectMutation,
  ]) {
    if (sideEffectMutation != null) {
      var isCanceled = false;
      sideEffectMutation(() => isCanceled = true);
      if (isCanceled) return;
    }

    component.forceUpdate();
  }
```

So maybe the solution would be something like this. Do you agree?:

```dart
abstract class RearchComponent extends Component2 {

  [...]

  bool _needsBuild = false;

  @override
  ReactNode render() {
    // Clear dirty flag.
    _clearNeedsBuild();

    // Clears the old dependencies (which will be repopulated via WidgetHandle)
    clearDependencies();

    final componentHandle = _ComponentHandleImpl(
      _ComponentSideEffectApiProxyImpl(this),
      topLevelCapsuleContainer,
    );
    return build(componentHandle);
  }

  ///.
  ReactNode build(ComponentHandle use);

  /// Set dirty flag and schedule update on next event loop.
  void markNeedsBuild() {
    if (_needsBuild) {
      return;
    }

    _needsBuild = true;

    Future.microtask(forceUpdate);
  }

  void _clearNeedsBuild() {
    if (!_needsBuild) {
      return;
    }

    _needsBuild = false;
  }
}
```
```dart

[...]

@override
  void rebuild([
    void Function(void Function() cancelRebuild)? sideEffectMutation,
  ]) {
    if (sideEffectMutation != null) {
      var isCanceled = false;
      sideEffectMutation(() => isCanceled = true);
      if (isCanceled) return;
    }

    component.markNeedsBuild();
  }
```

**Reply by busslina** [reply]
And sorry if I'm boring you haha

**Reply by busslina** [reply]
Or maybe just calling`forceUpdate` is enough, as it maybe rebuilds once on multiple API `rebuild()` calls.
So I have to:
- Check if it's okay to call `foceUpdate` synchronously.
- Check if it's okay to call multiple times `forceUpdate` or I have to take care to only call it once.

**Reply by GregoryConrad** [reply]
Yes, that sounds about right

**Reply by busslina** [reply]
> Or maybe just calling `forceUpdate` is enough, as it maybe rebuilds once on multiple API `rebuild()` calls.

Yes, it's enough, no `dirty` logic or `async call` is needed:
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/4fe7583a-70ec-4358-b071-605ba23ec403)
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/a97fd552-2f02-42ab-bf1c-9962c004c0ea)

Now I have to care about using `props` (similar to Widget properties/`didUpdateWidget`) and it's done if nothing escapes me :)

**Reply by busslina** [reply]
After I test it in a real project I will send you a ZIP via Discord DM so you can modify it to your fits and integrate in your lib if you want it. I don't care about putting my name on it haha

**Reply by GregoryConrad** [reply]
You could make a PR if you wanted and then you can get the open source credit 😝

**Reply by busslina** [reply]
Okay, just that I don't document it well (neither commit messages or your coding commentary standards haha, which I like: I discovered many useful lints by digging on your code) so I will encourage you to change anything (code + comments) to fit those standards..
